random/sample.ts

```diff
@@ -0,0 +1,102 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+// This module is browser compatible.
+
+import {
+  defaultOptions as defaultRandomOptions,
+  type RandomOptions,
+} from "./_types.ts";
+import { randomIntegerBetween } from "./integer_between.ts";
+
+/**
+ * Options for {@linkcode sample}.
+ */
+export type SampleOptions = RandomOptions & {
+  /**
+   * An array of weights corresponding to each item in the input array.
+   * If supplied, this is used to determine the probability of each item being
+   * selected.
+   *
+   * @default {undefined}
+   */
+  weights?: readonly number[] | undefined;
+};
+
+const defaultOptions: SampleOptions = {
+  ...defaultRandomOptions,
+  weights: undefined,
+};
+
+/**
+ * Returns a random element from the given array.
+ *
+ * @typeParam T The type of the elements in the array.
+ * @typeParam O The type of the accumulator.
+ *
+ * @param array The array to sample from.
+ * @param options Options modifying the sampling behavior.
+ *
+ * @returns A random element from the given array, or `undefined` if the array
+ * is empty.
+ *
+ * @example Basic usage
+ * ```ts
+ * import { sample } from "@std/random/sample";
+ * import { assertArrayIncludes } from "@std/assert";
+ *
+ * const numbers = [1, 2, 3, 4];
+ * const sampled = sample(numbers);
+ *
+ * assertArrayIncludes(numbers, [sampled]);
+ * ```
+ *
+ * @example Using `weights` option
+ * ```ts no-assert
+ * import { sample } from "@std/random/sample";
+ *
+ * const weighted = new Map([["a", 5], ["b", 3], ["c", 2]]);
+ * const result = sample([...weighted.keys()], { weights: [...weighted.values()] });
+ * // gives "a" 50% of the time, "b" 30% of the time, and "c" 20% of the time
```

Comment:
Can we make this a little easier on the eyes by simply having two arrays; `array` and `weighted`? The comment should make this snippet understandable.

Reply:
I wanted the example to use "zipped" values as a starting point, then only "unzip" them on calling the function, as I think starting with unzipped values isn't a best practice and leads to brittle code.

E.g.
```ts
const items = [
    'some long string that causes the code formatter to line wrap',
    'something else',
    'third thing',
];
const weights = [5, 3, 2];
```

It's a lot less easy to see which weight goes with which value, harder to refactor, and easier to make mistakes (e.g. pushing to `items` without pushing to `weights`, calling `sort` on one or both arrays causing them to get out-of-sync, etc).

Reply:
I still think simple arrays incur less cognitive load. Not a strong opinion so happy to defer to your decision.

Reply:
On second thoughts I think you're right. Changed the example 👍